public/blog/2025-01-bio-foundation-models/2025-01-bio-fountation-models.md

```diff
@@ -0,0 +1,84 @@
+
+We will now use a protein foundation model from Meta called ESM-2. Visit the Hugging Face website to inspect the model. ESM-2 comes in different sizes, from 8M to 15B parameters. We will use the smallest version to save computation time, but feel free to upgrade the model. To find foundation models in general, you can search through Hugging Face. Copy the model’s name on the top of the page (e.g., ‘facebook/esm2_t6_8M_UR50D’) and return to Orange.
+
+Open a Python Script widget and copy the code below inside the python_script function. Let’s quickly go through what it does. First, it imports the necessary packages like _numpy_ and _torch_. Then, we fetch sequences from our in_data, which is a DataTable on the input of the Python Script widget. The next three lines are where we define our foundation model, in our case, the "facebook/esm2_t6_8M_UR50D". Tokenizer is used to transform our sequence into tokens, the language of the model. Lastly, we define the model itself with the EsmModel class. Note that if you use another foundation model (e.g., DNABert), you only need to change the variable ‘model_name.’ In the following steps, we iterate through sequences, using the tokenizer and the model to extract the protein embedding on the output. The last for loop goes through protein embeddings and appends each column to our data table. Finally, we output the resulting table with protein embeddings as columns.
+
+```python
+# import necessary packages and classes
+# Feel free to use ChatGPT to explain this code in furher detail
+import numpy as np
+import torch
+from Orange.data import ContinuousVariable
+from transformers import AutoTokenizer, AutoModel
+
+# we extract the column sequences from our data table
+sequences = in_data.get_column("sequence")
+
+# we define model_name, while AutoTokenizer and EsmModel download the foundation model
+model_name = "facebook/esm2_t6_8M_UR50D"
+tokenizer = AutoTokenizer.from_pretrained(model_name)
+model = AutoModel.from_pretrained(model_name)
+
+embeddings = []
+for i in range(len(sequences)):
+    # First, we have to transform each sequence into tokens
+    inputs = tokenizer(sequences[i], return_tensors="pt")
+  
+    with torch.no_grad():
+        # Then, we calculate the embeddings of the sequence with ESM-2 model
+        outputs = model(**inputs)
+    # Importantly, we use .detach() to reduce the memory load
+    embeddings.append(outputs.last_hidden_state[0][0].detach().numpy())
+embeddings = np.array(embeddings)
+
+EMB_DIM = embeddings.shape[-1]
+for i in range(EMB_DIM):
+    # Lastly, we append each column of embeddings to our data as a Continuous variable
+    in_data = in_data.add_column(
```

Comment:
Ne spreminjaj `in_data` (ni lepa praksa, da widgeti spreminjajo svoje inpute). Na zacetku naredi lepo out_data.copy(), in dodaj tam.